### PR TITLE
[Testcase RefactoringEnable tensor ops tests for out-of-tree accelerators via capability gates

### PR DESCRIPTION
**Pull request by youlituao**
## Summary

`test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py` tests
ShardedTensor operations (deepcopy, inplace_copy, clone, detach,
set_requires_grad) across 4 distributed ranks. The test class uses
`@requires_accelerator_dist_backend(["nccl", "xccl"])`, which prevents
out-of-tree accelerators (PrivateUse1-based backends) from running these
tests because the hardcoded backend list does not include `privateuse1`.

This PR keeps `ShardedTensorTestBase` + `with_comms` + `TEST_GPU_NUM`
intact and makes minimal changes to enable out-of-tree accelerator
support:
- Replaces `@requires_accelerator_dist_backend(["nccl", "xccl"])` with
  `@requires_capabilities(Capability.distributed.backend)`, which gates
  on device-declared capability rather than a hardcoded backend name list.
- Adds `hw_classification = HardwareClassification.ACCELERATOR`.
- Adds `instantiate_device_type_tests(TestTensorOps, globals(),
  except_for="cpu", allow_xpu=True)` to generate per-accelerator test
  variants while excluding CPU. `allow_xpu=True` preserves the original
  XPU coverage (the original `["nccl", "xccl"]` list included `xccl`).
- Adds `device` parameter to all 5 test methods. The `device` parameter
  is injected by `instantiate_device_type_tests`; test methods extract
  the bare device type via `torch.device(device).type` for
  `ChunkShardingSpec` placements.
- Removes module-level `device_type`/`backend` globals. `@with_comms(init_rpc=False)`
  relies on `ShardedTensorTestBase.backend` property (PR #194992) for
  dynamic backend resolution per device variant.

## Changes

- Replaced `@requires_accelerator_dist_backend(["nccl", "xccl"])` with
  `@requires_capabilities(Capability.distributed.backend)` on all 5
  test methods.
- Added `hw_classification = HardwareClassification.ACCELERATOR` class
  attribute.
- Added `instantiate_device_type_tests(TestTensorOps, globals(),
  except_for="cpu", allow_xpu=True)` at file end.
- Added `device` parameter to all 5 test method signatures:
  `def test_xxx(self)` → `def test_xxx(self, device)`.
- Added `device_type = torch.device(device).type` at the start of each
  test method body to extract the bare device type from the injected
  `device` parameter (e.g., `"npu:0"` → `"npu"`).
- Removed module-level `device_type` and `backend` globals.
- Changed `@with_comms(init_rpc=False, backend=backend)` to
  `@with_comms(init_rpc=False)` on all 5 test methods.
- Added imports: `Capability`, `instantiate_device_type_tests`,
  `requires_capabilities` from `common_device_type`;
  `HardwareClassification` from `common_utils`.
- Removed import: `requires_accelerator_dist_backend` from
  `common_distributed`.

## Dependencies

This PR relies on the following upstream PRs (submitted but not yet merged):

| Infrastructure | Source PR | Status |
|---|---|---|
| `ShardedTensorTestBase` backend property + `with_comms(backend=None)` + `init_pg` device binding | #194992 | Pending merge |
| `Capability` class + `requires_capabilities` decorator | #191919 | Pending merge |
| `PrivateUse1TestBase._capabilities()` override | #193080 | Pending merge (depends on #191919) |
| `skip_if_lt_x_gpu` hardware-agnostic | #187650 | Pending merge |

Without these, non-CUDA device variants are either skipped or fail at
process group initialization. The `@requires_capabilities` decorator
degrades to skipped (not failed) when `#193080` is not applied, so this
PR can be merged before #193080 without risk.

## Not changed

- Base class remains `ShardedTensorTestBase`.
- `with_comms(init_rpc=False)` decorator retained on all test methods
  (backend resolved dynamically via `ShardedTensorTestBase.backend`
  property, no longer passed as a module-level global).
- `TEST_GPU_NUM` retained for `@skip_if_lt_x_gpu(TEST_GPU_NUM)`.
- All 5 test method bodies and assertions are unchanged (only the
  `device` parameter and `device_type` derivation added).
- `ChunkShardingSpec` configuration (dim=0, 4 placements, shape (12, 5))
  is unchanged.
- Test count is preserved (5 tests).

## Test plan

- `python test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py`
  on CUDA (A100, PyTorch 2.13.0a0, CUDA 12.8, 8 GPUs):
  `OK` — all 5 tests passed in 47.675s
  (`test_deep_copy_cuda`, `test_inplace_copy_cuda`, `test_clone_cuda`,
  `test_detach_cuda`, `test_set_requires_grad_cuda`).
  This confirms the refactoring does not break existing CUDA behavior.
- `python test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py`
  on an out-of-tree accelerator backend (PrivateUse1, torch_npu
  2.13.0+git45fbeae on PyTorch 2.13.0a0+gitfad7424, Ascend 910B3, 4 NPUs,
  with #194992, #191919, #193080, #187650 applied):
  `OK` — all 5 tests passed in 89.379s
  (`test_deep_copy_npu`, `test_inplace_copy_npu`, `test_clone_npu`,
  `test_detach_npu`, `test_set_requires_grad_npu`).
- CPU: no variants generated (`except_for="cpu"`).
- `lintrunner`: all 62 checks passed (PYFMT/FLAKE8/RUFF/CODESPELL etc.).
- `ruff check`: All checks passed.

### CODE REVIEW

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py [hunk 1]
```diff
@@ -5,29 +5,29 @@
...
+    world_size = 4
 
+    @classmethod
+    def backend_str(cls) -> str:
+        acc = torch.accelerator.current_accelerator()
+        dt = acc.type if acc is not None else "cpu"
+        return torch.distributed.get_default_backend_for_device(dt)
 
-class TestTensorOps(ShardedTensorTestBase):
-    @with_comms(init_rpc=False, backend=backend)
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:27]: should not delete with_comms

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py [hunk 2]
```diff
@@ -5,29 +5,29 @@
...
-    @skip_if_lt_x_gpu(TEST_GPU_NUM)
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:28]: keep TEST_GPU_NUM, should not modify to 4

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py [hunk 3]
```diff
@@ -5,29 +5,29 @@
...
-    @requires_accelerator_dist_backend(["nccl", "xccl"])
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:29]: requires_accelerator_dist_backend, uggest using the capability registration mechanism #191919

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py [hunk 4]
```diff
@@ -5,29 +5,29 @@
...
-    TEST_GPU_NUM,
-    with_comms,
-)
 
 
-device_type = (
-    acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
-)
-backend = torch.distributed.get_default_backend_for_device(device_type)
+class TestTensorOps(MultiProcContinuousTest):
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py]: missing hw_classification

**Review by JamesD18 — COMMENTED** [review]
instantiate_device_type_tests(TestTensorOps, globals()) CPU needs to be excluded.

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py (hunk 1 above)
Comment (youlituao) [reply test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:27]: Acknowledged. with_comms(init_rpc=False, backend=backend) is retained on all 5 test methods. The ShardedTensorTestBase base class is also kept. The module-level backend global is resolved at import time via torch.distributed.get_default_backend_for_device(device_type), consistent with the other ShardedTensor test files in the same directory.

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py (hunk 4 above)
Comment (youlituao) [reply test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py]: Acknowledged. hw_classification = HardwareClassification.ACCELERATOR is added to TestTensorOps. This matches the instantiate_device_type_tests(except_for="cpu") mechanism — the class generates per-accelerator variants (CUDA, PrivateUse1, etc.) while excluding CPU, so ACCELERATOR is the correct classification.

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py (hunk 2 above)
Comment (youlituao) [reply test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:28]: Acknowledged. TEST_GPU_NUM is retained and @skip_if_lt_x_gpu(TEST_GPU_NUM) is unchanged on all test methods. No hardcoded world_size = 4 or @skip_if_lt_x_gpu(4).

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py (hunk 3 above)
Comment (youlituao) [reply test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py:29]: Acknowledged. @requires_accelerator_dist_backend(["nccl", "xccl"]) is replaced with @requires_capabilities(Capability.distributed.backend) on all 5 test methods, using the capability registration mechanism from PR #191919. This gates on device-declared capability rather than a hardcoded backend name list, so out-of-tree accelerators (PrivateUse1-based backends) are automatically supported without modifying the check. PrivateUse1TestBase._capabilities() override is provided by PR #193080. When #193080 is not applied, PrivateUse1 variants are correctly skipped (skipped, not failed).

**Comment by youlituao** [comment]
> instantiate_device_type_tests(TestTensorOps, globals()) CPU needs to be excluded.

Acknowledged. instantiate_device_type_tests(TestTensorOps, globals(), except_for="cpu") is used instead of instantiate_device_type_tests(TestTensorOps, globals()). CPU variants are not generated, since these tests require a distributed accelerator backend (@skip_if_lt_x_gpu(TEST_GPU_NUM) + @requires_capabilities(Capability.distributed.backend)).

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py [hunk 5]
```diff
@@ -24,9 +26,11 @@
 
 
 class TestTensorOps(ShardedTensorTestBase):
+    hw_classification = HardwareClassification.ACCELERATOR
+
     @with_comms(init_rpc=False, backend=backend)
     @skip_if_lt_x_gpu(TEST_GPU_NUM)
-    @requires_accelerator_dist_backend(["nccl", "xccl"])
+    @requires_capabilities(Capability.distributed.backend)
     def test_deep_copy(self):
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py]: The device parameter is missing. Corresponding modifications are also required for other methods.

**Review by JamesD18 — APPROVED** [review]
LGTM.Thanks for addressing the review comments.

File: test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py (hunk 5 above)
Comment (youlituao) [reply test/distributed/_shard/sharded_tensor/ops/test_tensor_ops.py]: Addressed. The `device` parameter has been added to all 5 test methods:

```python
@with_comms(init_rpc=False, backend=backend)
@skip_if_lt_x_gpu(TEST_GPU_NUM)
@requires_capabilities(Capability.distributed.backend)
def test_deep_copy(self, device):
    device_type = torch.device(device).type
    ...
The device parameter is injected by instantiate_device_type_tests. Inside each method, torch.device(device).type extracts the bare device type (e.g., "npu:0" → "npu") for ChunkShardingSpec placements. ```
